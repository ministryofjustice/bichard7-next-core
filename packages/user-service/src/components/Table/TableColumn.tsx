import { cloneElement } from "react"
import KeyValuePair from "types/KeyValuePair"

interface Props {
  component?: JSX.Element
  item: KeyValuePair<string, string>
  field: string
}

const TableColumn = ({ component, item, field }: Props) => {
  return <td className="govuk-table__cell">{component ? cloneElement(component, { item }) : item[field]}</td>
}

export default TableColumn
