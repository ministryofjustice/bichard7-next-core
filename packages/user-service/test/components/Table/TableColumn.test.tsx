import TableColumn from "components/Table/TableColumn"
import { render } from "@testing-library/react"
import { ReactNode } from "react"
import KeyValuePair from "types/KeyValuePair"

it("should render the field value when custom column component is not provided", () => {
  const item = { dummyField: "DummyValue" }

  const WrapperComponent = ({ children }: { children: ReactNode }) => (
    <table>
      <tbody>
        <tr>{children}</tr>
      </tbody>
    </table>
  )
  const { container } = render(<TableColumn field="dummyField" item={item} />, {
    wrapper: WrapperComponent as React.ComponentType
  })

  expect(container).toMatchSnapshot()
})

it("should render the component when custom column component is provided", () => {
  const dummyItem = { dummyField: "DummyValue" }

  type ColumnComponentProps = { item?: KeyValuePair<string, string>; field: string }
  type WrapperProps = { children: ReactNode }

  const ColumnComponent = ({ item, field }: ColumnComponentProps) => <h1>{item?.[field]}</h1>

  const WrapperComponent = ({ children }: WrapperProps) => (
    <table>
      <tbody>
        <tr>{children}</tr>
      </tbody>
    </table>
  )
  const { container } = render(
    <TableColumn field="dummyField" item={dummyItem} component={<ColumnComponent field="dummyField" />} />,
    {
      wrapper: WrapperComponent as React.ComponentType
    }
  )

  expect(container).toMatchSnapshot()
})
