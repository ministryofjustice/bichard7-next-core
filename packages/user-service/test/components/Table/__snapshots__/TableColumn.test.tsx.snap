// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should render the component when custom column component is provided 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="govuk-table__cell"
        >
          <h1>
            DummyValue
          </h1>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should render the field value when custom column component is not provided 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="govuk-table__cell"
        >
          DummyValue
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
